docs/package.mill

```diff
@@ -7,9 +7,70 @@ import scala.jdk.CollectionConverters._
 
 /** Generates the mill documentation with Antora. */
 object `package` extends RootModule {
+  trait CustomUnidocModule extends UnidocModule {
```

Comment:
I added that to workaround issues apparently caused by [coursier/dependency](https://github.com/coursier/dependency) being around. I agree it'd be better to address the underlying cause. I couldn't make sense of the issue and don't know that much about unidoc, so I went for that workaround first.

The workaround tries to remove coursier/dependency from the class path passed to unidoc. It is two-fold: short-term workaround and slightly longer-term one.

To tweak the class path passed to unidoc and make the CI green right now, we need to change `UnidocModule#unidocCommon`, so that it allows users to tweak the class path passed to unidoc. `CustomUnidocModule` copies that task source, and exposes `unidocCompileClasspath`, that is customized in `object site`.

In the slightly longer term, the PR here changes the original `UnidocModule`, so that we don't need to copy `unidocCommon` sources to adjust the unidoc class path. So that as soon as Mill is rebootstrapped (with the PR here), `CustomUnidocModule` can be removed, as the original `UnidocModule` now allows the class path customization.

Reply:
TL/DR is that `CustomUnidocModule` is meant to be removed as soon as Mill can be rebootstrapped. Fixing the original unidoc issue would be better, but I couldn't make sense of it, so I went for a workaround first.

Reply:
Seems we're running into https://github.com/scala/bug/issues/10028

Reply:
This seems specific to Scala 2. Scala CLI depends on coursier/dependency too, and doesn't have that problem in [its scaladoc](https://javadoc.io/doc/org.virtuslab.scala-cli/cli_3/1.5.1/index.html), while coursier, still on Scala 2, [has](https://javadoc.io/doc/io.get-coursier/coursier_2.13/2.1.18/index.html).

Reply:
@alexarchambault if the problem is just broken links, rather than patching the unidoc code in so much detail how about we just `rm -rf` the relevant folder after unidoc generation. That will accomplish the same thing and be much easier to understand for readers and delete when it's no longer necessary

Reply:
So I removed the `CustomUnidocModule` stuff, but this PR now needs the Mill build to be rebootstrapped on top of https://github.com/com-lihaoyi/mill/pull/4002

Reply:
> if the problem is just broken links, rather than patching the unidoc code in so much detail how about we just `rm -rf` the relevant folder after unidoc generation. That will accomplish the same thing and be much easier to understand for readers and delete when it's no longer necessary

You mean, not check for broken links for parts of the site? I'm not sure we can post-process the generated site that easily, while not leaving broken links behind.

Reply:
Yes that's right. If we can't post-process the generated site, we can just patch the broken links checker to ignore the links in particular files or folders

https://github.com/com-lihaoyi/mill/blob/4729afc9493b161110587f9518542138dca58c0e/docs/package.mill#L453

Reply:
That would make the CI green, but excluding coursier/dependency from the class path passed to scaladoc has the advantage of making the generated API site cleaner. The scaladoc issue adds a `dependency` root package in the Mill scaladoc, like this:
<img width="1197" alt="Capture d’écran 2024-11-21 à 13 14 12" src="https://github.com/user-attachments/assets/e1daded7-80c5-4480-aba7-932baacce89c">

This `dependency` root package goes away when excluding the coursier/dependency JAR.

Reply:
got it, let's go with the other PR and re-bootstrap it then

Reply:
Merged the other PR and kicked off a publish on latest main branch https://github.com/com-lihaoyi/mill/actions/runs/11953290776, once that's out you can use the unstable version in this PR

Reply:
Seems it worked